lpc55_sign/src/debug_auth.rs

```diff
@@ -0,0 +1,213 @@
+        let socc = u32::from_le_bytes(value[0x4..0x8].try_into().unwrap());
+
+        let uuid = <[u8; 16]>::try_from(&value[0x8..0x18]).unwrap();
+
+        let rotk_revoke = u32::from_le_bytes(value[0x18..0x1C].try_into().unwrap());
+        let rotk_revoke = [
+            rotk_revoke & (1 << 0) != 0,
+            rotk_revoke & (1 << 1) != 0,
+            rotk_revoke & (1 << 2) != 0,
+            rotk_revoke & (1 << 3) != 0,
+        ];
+
+        let rotkh = <[u8; 32]>::try_from(&value[0x1C..0x3C]).unwrap();
+
+        let cc_socu_pin = u32::from_le_bytes(value[0x3C..0x40].try_into().unwrap());
+        let cc_socu_pin = CCSOCUPin(cc_socu_pin);
+
+        let cc_socu_dflt = u32::from_le_bytes(value[0x40..0x44].try_into().unwrap());
+        let cc_socu_dflt = CCSOCUDflt(cc_socu_dflt);
+
+        let vendor_usage = u32::from_le_bytes(value[0x44..0x48].try_into().unwrap());
+
+        let challenge_vector = <[u8; 32]>::try_from(&value[0x48..]).unwrap();
+
+        Ok(Self {
+            version_major,
+            version_minor,
+            socc,
+            uuid,
+            rotk_revoke,
+            rotkh,
+            cc_socu_pin,
+            cc_socu_dflt,
+            vendor_usage,
+            challenge_vector,
+        })
+    }
+}
+
+pub fn debug_credential(
+    root_certs: Vec<Certificate>,
+    root_key: &RsaPrivateKey,
+    debug_key: &RsaPublicKey,
+    uuid: &[u8; 16],
+    vendor_usage: u32,
+    debug_settings: DebugSettings,
+    beacon: u16,
+) -> Result<Vec<u8>, Error> {
+    let root_public_key = root_key.to_public_key();
+
+    root_certs
+        .iter()
+        .fold(Err(Error::RootNotFound), |acc, cert| {
+            match public_key(cert) {
+                Err(error) => Err(error.into()),
+                Ok(key) => {
+                    if key == root_public_key {
+                        Ok(())
+                    } else {
+                        acc
+                    }
+                }
+            }
+        })?;
+
+    let debug_key_size_bits = debug_key.size() * 8;
+    let root_key_size_bits = root_key.size() * 8;
+    if root_key_size_bits != debug_key_size_bits {
+        return Err(Error::VaryingPublicKeySizes);
+    };
+    let (version_major, version_minor): (u16, u16) = match root_key_size_bits {
+        2048 => (1, 0),
+        4096 => (1, 1),
+        _ => {
+            return Err(Error::UnsupportedRsaKeySize {
+                key_size: root_key_size_bits,
+            })
+        }
+    };
+
+    let mut dc_bytes: Vec<u8> = Vec::new();
+    // NXP UM11126 claims this is a single little-endian u32 version field.
+    // NXP's spsdk tooling implements it as done here: two little-endian u16s
+    // with version_major at the lower offset.  See
+    // https://github.com/nxp-mcuxpresso/spsdk/blob/5da31d96a020bd65e5834ea5ac1b68327ea965ef/spsdk/dat/debug_credential.py#L728.
+    dc_bytes.extend_from_slice(&version_major.to_le_bytes());
```

Comment:
(I'd also encourage implementing debug credential using a zerocopy struct instead of hand serializing. Happy to help with the details.)